truffleruby/src/main/java/org/truffleruby/core/thread/ThreadManager.java

```diff
@@ -216,20 +221,47 @@ public DynamicObject getRootThread() {
      */
     @TruffleBoundary
     public <T> T runUntilResult(Node currentNode, BlockingAction<T> action) {
+        return runUntilResult(currentNode, action, null);
+    }
+
+    /**
+     * Runs {@code action} until it returns a non-null value.
+     * The blocking action might be {@link Thread#interrupted()}, for instance by
+     * the {@link SafepointManager}, in which case it will be run again.
+     * The unblocking action is registered with the thread manager and will be invoked
+     * if the thread manager interrupts the thread. If the blocking action is making a
+     * native call, simply interrupting the thread will not unblock the action. It is the
+     * responsibility of the unblocking action to break out of the native call so the thread
+     * can be interrupted.
+     *
+     * @param blockingAction must not touch any Ruby state
+     * @param unblockingAction must not touch any Ruby state
+     * @return the first non-null return value from {@code action}
+     */
+    @TruffleBoundary
+    public <T> T runUntilResult(Node currentNode, BlockingAction<T> blockingAction, UnblockingAction unblockingAction) {
         final DynamicObject runningThread = getCurrentThread();
         T result = null;
 
+        if (unblockingAction != null) {
+            unblockingActions.put(Layouts.THREAD.getThread(runningThread), unblockingAction);
+        }
+
         do {
             Layouts.THREAD.setStatus(runningThread, ThreadStatus.SLEEP);
 
             try {
                 try {
-                    result = action.block();
+                    result = blockingAction.block();
                 } finally {
                     Layouts.THREAD.setStatus(runningThread, ThreadStatus.RUN);
                 }
             } catch (InterruptedException e) {
                 // We were interrupted, possibly by the SafepointManager.
+                if (unblockingAction != null) {
+                    unblockingActions.put(Layouts.THREAD.getThread(runningThread), unblockingAction);
+                }
+
                 context.getSafepointManager().pollFromBlockingCall(currentNode);
             }
         } while (result == null);
```

Comment:
I think we should remove it from `unblockingActions` here, so it's not called after a blocking call is finished.
We might as well make this re-entrant and restore the unblocking action fro this thread before it was changed.